File: 0098/dfs.cpp
#include <gtest/gtest.h>

#include <cstdlib>
#include <numeric>
#include <string>
#include <vector>

#include "leetcode/tree.hpp"

using namespace std;
using namespace LeetCode;

// used implmentation in https://leetcode.com/problems/validate-binary-search-tree/discuss/32141/C%2B%2B-simple-recursive-solution
// No need to use buggy INT_MIN and INT_MAX
class Solution {
 public:
  bool isValidBST(TreeNode* root) {
    return dfs(root, nullptr, nullptr);
  }

  bool dfs(TreeNode* node, TreeNode* min_node, TreeNode* max_node) {
    if (!node) { return true; }
    if (min_node && node->val <= min_node->val) { return false; }
    if (max_node && max_node->val <= node->val) { return false; }
    if (!dfs(node->left, min_node, node)) { return false; }
    if (!dfs(node->right, node, max_node)) { return false; }
    return true;
  }
};

int main() {
  Solution sol;
  TreeNode* root;
  bool expected;

  root = CreateTreeNode("2 1 3");
  expected = true;
  EXPECT_EQ(expected, sol.isValidBST(root));
  DeleteTreeNode(root);

  root = CreateTreeNode("5 1 4 # # 3 6");
  expected = false;
  EXPECT_EQ(expected, sol.isValidBST(root));
  DeleteTreeNode(root);

  root = CreateTreeNode("1 1");
  expected = false;
  EXPECT_EQ(expected, sol.isValidBST(root));
  DeleteTreeNode(root);

  root = CreateTreeNode("2147483647");
  expected = true;
  EXPECT_EQ(expected, sol.isValidBST(root));
  DeleteTreeNode(root);

  root = CreateTreeNode("-2147483648 -2147483648");
  expected = false;
  EXPECT_EQ(expected, sol.isValidBST(root));
  DeleteTreeNode(root);

  root = CreateTreeNode("-2147483648 # 2147483647");
  expected = true;
  EXPECT_EQ(expected, sol.isValidBST(root));
  DeleteTreeNode(root);

  exit(EXIT_SUCCESS);
}

File: 0101/dfs.rs
use std::rc::Rc;
use std::cell::RefCell;

use leetcode::{TreeNode, create_treenode};

struct Solution;

impl Solution {
    pub fn is_symmetric(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        if root.is_none() {
            return false;
        }
        let root_ref = root.as_ref().unwrap().borrow();
        return Solution::recursive(&root_ref.left, &root_ref.right);
    }

    fn recursive(left: &Option<Rc<RefCell<TreeNode>>>, right: &Option<Rc<RefCell<TreeNode>>>) -> bool {
        if (left.is_none() && right.is_some()) || (left.is_some() && right.is_none()) {
            return false;
        } else if left.is_none() && right.is_none() {
            return true;
        }
        let (left_ref, right_ref) = (left.as_ref().unwrap().borrow(), right.as_ref().unwrap().borrow());
        if left_ref.val != right_ref.val {
            return false;
        }
        return Solution::recursive(&left_ref.left, &right_ref.right) && Solution::recursive(&left_ref.right, &right_ref.left);
    }
}

fn main() {
    let (root, expected) = (create_treenode("1 2 2 3 4 4 3"), true);
    assert_eq!(Solution::is_symmetric(root), expected);

    let (root, expected) = (create_treenode("1 2 2 # 3 # 3"), false);
    assert_eq!(Solution::is_symmetric(root), expected);

    let (root, expected) = (create_treenode("2 3 3 4 5 5"), false);
    assert_eq!(Solution::is_symmetric(root), expected);
}
